Mathlib/RingTheory/Bialgebra.lean

```diff
@@ -0,0 +1,83 @@
+/-
+Copyright (c) 2024 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.Coalgebra
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Bialgebras
+
+In this file we define `Bialgebra`, and provide instances for:
+
+* Commutative semirings: `CommSemiring.toBialgebra`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Bialgebra>
+-/
+
+suppress_compilation
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A bialgebra over a commutative (semi)ring `R` is both an algebra and a coalgebra over `R`, such
+that the counit and comultiplication are algebra morphisms. -/
+class Bialgebra (R : Type u) (A : Type v) [CommSemiring R] [Semiring A] extends
+    Algebra R A, Coalgebra R A where
+  -- The counit is an algebra morphism
+  /-- The counit on a bialgebra preserves 1. -/
+  counit_one : counit 1 = 1
+  /-- The counit on a bialgebra preserves multiplication. -/
+  counit_mul : ∀ a₁ a₂ : A, counit (a₁ * a₂) = counit a₁ * counit a₂
+  -- The comultiplication is an algebra morphism
+  /-- The comultiplication on a bialgebra preserves `1`. -/
+  comul_one : comul 1 = 1
+  /-- The comultiplication on a bialgebra preserves multiplication. -/
+  comul_mul : ∀ a₁ a₂ : A, comul (a₁ * a₂) = comul a₁ * comul a₂
+
+namespace Bialgebra
+variable {R : Type u} {A : Type v}
```

Comment:
What about making `R` and `A` explicit here, so one can use `counitAlgHom R A` (similarly to `algebraMap R A`) without knowing the name of the instance?